lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -29,4 +98,276 @@
+    new network and route requests by integrated units,
+    * Synchronize the databags of all requiring units with the router table of the
+    provider charm
+
+    Attributes:
+        charm:
+            The Charm object that instantiates this class
+        _stored:
+            The persistent state that keeps the internal routing table.
+    """
+
+    on = RouterProviderCharmEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "ip-router")
+        self.charm = charm
+        self._stored.set_default(routing_table={})
+        self.framework.observe(
+            charm.on.ip_router_relation_changed, self._on_ip_router_relation_changed
+        )
+        self.framework.observe(
+            charm.on.ip_router_relation_joined, self._on_ip_router_relation_joined
+        )
+        self.framework.observe(
+            charm.on.ip_router_relation_departed, self._on_ip_router_relation_departed
+        )
+
+    def _on_ip_router_relation_joined(self, event: RelationJoinedEvent):
+        """When a new unit or app joins the relation, add its name to the routing table"""
+        self._stored.routing_table.update({event.relation.app.name: []})
+        self._sync_routing_tables()
+
+    def _on_ip_router_relation_changed(self, event: RelationChangedEvent):
+        """Update the internal routing table state to reflect changes in
+        requirer units' new network requests.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if "networks" not in event.relation.data[event.relation.app]:
+            return
+
+        new_network = event.relation.data[event.relation.app]["networks"]
+        if new_network == self._stored.routing_table[event.relation.app.name]:
+            return
+        self._stored.routing_table[event.relation.app.name] = new_network
+        self._sync_routing_tables()
+        self.on.routing_table_updated.emit()
+
+    def _on_ip_router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if len(event.relation.units) == 0:
+            self._stored.routing_table.pop(event.app.name)
+            self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Read-only way to get the current routing table"""
+        return deepcopy(self._stored.routing_table._under)  # RFC: Is there a better way?
+
+    def get_flattened_routing_table(self) -> List[Network]:
+        """Returns a read-only internal routing table that's flattened
+
+        Returns:
+            A list of objects of type `Network`
+        """
+        internal_rt = self.get_routing_table()
+        final_rt = []
+        for networks in internal_rt.values():
+            if type(networks) is not str:
+                continue
+            for network in json.loads(networks):
```

Comment:
Can this raise? I do not see validation from the data received in the Provider. The Requirer side is not guaranteed to be using the library to use this relation, so we cannot rely only on that side doing the validation.

Reply:
1) Even if there's an error, it should be raised when the databags are being updated and not here, this should return something always.

2) I'm still conflicted on the approach to validating on the provider side. My goal with this library was to make sure that an invalid state is unreachable with proper library use. 

If we start to build for people using the library in an unintended way, we open a pandora's box of problems. What if a user decides to put random data in the provider for example? Should we handle that everywhere as well? Since the provider can't actually do anything about the requirer's databag, there's no way to save the relationship from the provider side. A notification must be sent to the requirer that made an error, so raising on the provider side is pointless in that regard. On top of that, if we raise an exception here, would that not break other, working `ip-router` relations?

The best alternative option imo is to just ignore the erroneous databag completely and log an error, but if a user was using the library and not seeing their network in the provider it's difficult for them to notice a single error log, if they don't use the provided `request_network` function. 

I'd love to hear your opinions on this.

Reply:
My concern is that if anything in `networks` is not JSON de-serializable, it will raise an exception that will crash the charm, putting it in an error state that will be difficult for an admin to recover from.

I think at least handling the exception here and logging it would be better, as the charm would continue to work, even if one requirer put garbage in.

Reply:
So, option 2. Done.